Implementation of mempool synchronization

bissias opened the pull request:
This PR provides an implementation of mempool synchronization using Graphene set reconciliation primitives. The overall approach is for peers to pull synchronization objects from other peers at regular intervals (currently once every 30 seconds). Each interval, the requesting peer randomly chooses a peer from whom to request synchronization. When a responding peer receives a mempool sync request, he makes a best effort to send to the requester all transactions in his mempool that are missing from the requester's mempool. Synchronization is not necessarily symmetric: peer `A` can sync with the mempool of `B`, but `B` need not necessarily sync with `A`. 

For DoS prevention, each peer will accept a synchronization request from any other peer at most once every 25 seconds. For any given synchronization session that is in-flight, the responding peer will accept exactly one request for missing transactions. Similarly, each requesting peer will accept exactly one missing transaction response per synchronization request. Any messages exceeding the above limits will result in an increase in the ban counter for the offending peer. 

Validation includes rpc and manual testing. The rpc test sets up 3 nodes and turns off normal transaction propagation between them. Each peer generates a batch of transactions and conducts mempool sync with its neighbors. The test passes if all mempools contain all transactions after 180 seconds or less. For manual testing, @sickpig setup an edge node on mainnet running mempool sync. On my machine locally, I setup an internal node also running mempool sync and connecting exclusively to the edge node. Normal transaction propagation was disabled between edge and internal nodes. Attached is a plot of the transaction counts in the mempools of each node over a 1 day period. The lines are nearly identical, indicating close synchronization between the two mempools. 
<img width="579" alt="sync_test" src="https://user-images.githubusercontent.com/1005112/65324924-8926ea00-db7b-11e9-90b1-e3577563c684.png">

bissias committed:
basic implementation of mempool sync

ptschip commented on qa/rpc-tests/mempoolsync.py:
```diff
@@ -0,0 +1,118 @@
+            self.nodes[2].sendtoaddress(self.nodes[2].getnewaddress(), Decimal("10"))
+
+        time.sleep(5)
+
```
generally not good to use sleeps, better to use WaitFor() to know that you got the result you're looking for.

bissias replied:
Done.

Greg-Griffith commented:
I verbally gave this feedback already, writing it here so it doesn't get lost. 

peers should exchange information about minimum fees and maximum mempool sizes to avoid requesting information that will be discard when received due to a configured mempool constraint.

bissias commented:
@Greg-Griffith can you think of a better way to calculate current mempool size (in bytes) than simply iterating though `CTxMemPool.mapTx`?

Greg-Griffith commented:
@bissias we already track this data so you shouldnt have to iterate through anything. 
this function returns the mempools size in bytes. (txmempool.h line 144 ish)
size_t DynamicMemoryUsage() const { return nUsageSize; }

Greg-Griffith commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -96,9 +96,9 @@ bool HandleMempoolSyncRequest(CDataStream &vRecv, CNode *pfrom)
-                uint64_t nSatoshiPerK = (uint64_t)(1000 * nFee / (double)nTxSize);
+                size_t nTxSize = it->GetTx().GetTxSize();
+                int64_t nFee = it->GetFee();
+                uint64_t nSatoshiPerK = (uint64_t)(1000 * nFee / (1.0 * nTxSize));
```
you should remove the 1.0 here. any decimals would be truncated by the cast to a uint64_t anyway

bissias replied:
Excellent point. Fixed.

Greg-Griffith commented on src/xversionkeys.dat:
```diff
@@ -32,6 +32,10 @@ KEY i BU_XTHIN_VERSION                          0x0002                   0x0003
+# Mempool sync preferences
+KEY i BU_MEMPOOL_SYNC                           0x0002                   0x0006         u64c
+KEY i BU_MEMPOOL_SYNC_MIN_VERSION_SUPPORTED     0x0002                   0x0007         u64c
+KEY i BU_MEMPOOL_SYNC_MAX_VERSION_SUPPORTED     0x0002                   0x0007         u64c
```
@bissias Is having both max and min version supported on 0x0007 a typo? I would think you would want different values for these keys.

bissias replied:
Good catch! Fixed.

Greg-Griffith commented on src/net.cpp:
```diff
@@ -2276,6 +2276,10 @@ void ThreadMessageHandler()
 
+        // randomly select node with whom to request mempoolsync
+        if (vNodesCopy.size() > 0)
+            requester.RequestMempoolSync(vNodesCopy[GetRandInt(vNodesCopy.size())]);
```
Imo it would be better to only attempt to sync mempools with nodes that are not in IBD. nodes in IBD will under most circumstances have a mempool that is far less complete than one that has been "online" for a longer period of time and trying to sync with them will be a wasted effort.

bissias replied:
That's a good point. What's a good way to check if they're in IBD?

Greg-Griffith commented on src/blockrelay/mempool_sync.h:
```diff
@@ -0,0 +1,187 @@
+#include <chrono>
+
+const uint64_t MEMPOOL_SYNC_MIN_VERSION_SUPPORTED = 0;
+const uint64_t MEMPOOL_SYNC_MAX_VERSION_SUPPORTED = 0;
```
please change these two constants to 
DEFAULT_MEMPOOL_SYNC_MIN_VERSION_SUPPORTED
DEFAULT_MEMPOOL_SYNC_MAX_VERSION_SUPPORTED
to match the naming convention of default values used by the rest of the codebase

bissias replied:
Done.

Greg-Griffith commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,404 @@
+    uint64_t lower = (uint64_t)std::max(peerMin, selfMin);
+
+    if (lower > upper)
+        throw std::runtime_error("Sender and receiver support incompatible mempool sync versions");
```
it might be useful here to instead of throwing an error to send some sort of reject message back to the requesting peer and add them to a "non negotiable" list for the duration of the connection so that we don't turn around and later request a sync from them just to have the same error happen. 

when a node receives the reject message it should add the peer that sent it to its own "non negotiable" list to avoid future requests to sync with that node and immediately try to negotiate a mempool sync with another peer since this sync failed.

bissias replied:
I'm going to paste here what we agreed upon offline. 

Before sending any mempool sync requests, a peer should: (1) check that the BU_MEMPOOL_SYNC xversion value is set to 1 and (2) check that its mempool sync version is compatible with the version of the would-be responder (again using xversion information).

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,444 @@
+                size_t nTxSize = it->GetTx().GetTxSize();
+                int64_t nFee = it->GetFee();
+                uint64_t nSatoshiPerK = (uint64_t)(1000 * nFee / nTxSize);
+
```
I think you should create a CFeeRate object here found in amount.h, to calculate nsatoshiperk,  then do a object.GetFeePerK().  Problem here is that what happens if nTxSize is zero, the CFeeRate class will already take care of that.

bissias replied:
Done.

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,444 @@
+
+    // Do not process unrequested memsynctx.
+    if (mempoolSyncRequested.count(pfrom) == 0)
+    {
```
where are the locks on cs_mempoolsync?  if they are implied before calling this function then add an AssertLockHeld(cs_mempoolsync) at the beginning of the method.

Same for all your other methods, i think there are a few more above here that I don't see any locking for.

bissias replied:
Good catch. Actually, the `mempoolSyncRequested` and `mempoolSyncResponded` structures were not being locked at all. The only thing that I used `cs_mempoolsync` for was to prevent multiple sync requests from being made simultaneously from ThreadHandler in net.cpp. I went ahead and changed that mutex to `cs_mempoolsync_init` and am now using `cs_mempoolsync` to lock the maps above in mempool_sync.cpp. I figured that it's best to keep the lock activated in net.cpp as lightweight as possible since it's in what I assume to be a potential bottleneck in the code.

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,444 @@
+        {
+            mempoolTxHashes.push_back(kv.first);
+        }
+    }
```
don't you have to add the commitQ hashes as well here? 

Also, why not use a set instead of a vector in case you get duplicates which could happen if for instance you remove the orphan pool lock and then the orphan slips into the mempool.

bissias replied:
Done. I also noticed that I wasn't checking CommitQ or orphanpool on the responder end. So I added that as well.

ptschip commented on src/blockrelay/mempool_sync.h:
```diff
@@ -0,0 +1,188 @@
+// frequency of synchronization (per peer) in milliseconds
+const int64_t MEMPOOLSYNC_FREQ_US = 30 * 1e6;
+const int64_t MEMPOOLSYNC_FREQ_GRACE_US = 5 * 1e6;
+// Use CVariableFastFilter if true, otherwise use CBloomFilter
```
don't you mean microseconds and not millis, otherwise would be a very long time

bissias replied:
Yes, another good catch. I did in fact mean microseconds.

ptschip commented on src/globals.cpp:
```diff
@@ -450,6 +475,10 @@ CThinBlockData thindata;
+std::map<CNode *, CMempoolSyncState> mempoolSyncRequested;
+std::map<CNode *, CMempoolSyncState> mempoolSyncResponded;
+std::chrono::time_point<std::chrono::high_resolution_clock> lastMempoolSync = std::chrono::high_resolution_clock::now();
+CCriticalSection cs_mempoolsync;
```
the map definition should include GUARDED_BY(cs_mempoolsync);  see other examples in globals.  Also you'll have then move CCriticalSection cs_mempoolsync to be before any guarded by statements.

bissias replied:
Done.

ptschip commented on src/requestManager.cpp:
```diff
@@ -1196,7 +1196,7 @@ void CRequestManager::FindNextBlocksToDownload(CNode *node, unsigned int count,
 {
-    LOCK(cs_mempoolsync);
+    LOCK(cs_mempoolsync_init);
 
```
I think you still have to use cs_mempoolsync here because of mempooSyncReqested is guareded under that lock.

bissias replied:
Ok, I see. Done.

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -287,17 +288,33 @@ bool CRequestMempoolSyncTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
+            txRef = CommitQGet(hash);
+            if (txRef != nullptr)
+                vTx.push_back(CommitQGet(hash));
 
```
-vTx.push_back(CommitQGet(hash));
+vTx.push_back(txRef);

bissias replied:
Stupid error, fixed.

gandrewstone commented on qa/rpc-tests/mempoolsync.py:
```diff
@@ -0,0 +1,64 @@
+
+        logging.info("Send 10 transactions from node2 (to its own address)")
+        for i in range(10):
+            self.nodes[2].sendtoaddress(self.nodes[2].getnewaddress(), Decimal("10"))
```
to speed the QA up, just get 1 address and send to it 10 times.

bissias replied:
Done.

gandrewstone commented on src/blockrelay/mempool_sync.h:
```diff
@@ -0,0 +1,188 @@
+class CMempoolSyncState
+{
+public:
+    std::chrono::time_point<std::chrono::high_resolution_clock> lastUpdated;
```
Use https://en.cppreference.com/w/cpp/chrono/steady_clock for all interval timing.  Use the others for recording and display of wall clock time only.  Otherwise the program can get screwed up when the wall clock is changed either manually or during daylight savings time.  Its ok to use std::chrono, especially if you envision this as being useful outside of bitcoind.  But if you want to be similar to the rest of the code and cut through all the namespaces, templates, and over-typing of std then just use the functions in utiltime.h:

/** Returns a monotonically increasing time for interval measurement (in nSec).  This number is unrelated to calendar time and is not affected by mock time during test */
uint64_t GetStopwatch();
/** Returns a monotonically increasing time for interval measurement (in uSec).  This number is unrelated to calendar time and is not affected by mock time during test */
inline uint64_t GetStopwatchMicros() { return GetStopwatch() / 1000; }

bissias replied:
Went with GetStopwatchMicros.

gandrewstone commented on src/blockrelay/mempool_sync.h:
```diff
@@ -0,0 +1,188 @@
+    }
+};
+
+extern std::map<CNode *, CMempoolSyncState> mempoolSyncRequested;
```
If you use a pointer as the key, and then dereference this pointer, you need to make sure that the CNode* pointer is not deleted.  CNode* pointers can be deleted whenever a node disconnects.  To do so, you'd need to call CNode::AddRef(), and hold that reference over the entire time CNode* pointers are stored in your map.  But if you do this, you'll prevent the cleanup of CNodes, so you need to implement code that periodically searches your maps for any disconnected CNodes, deletes the entry from your map and releases the ref.

Another possibility (which is more common in your situation) is to use the node id.  This id is a non-repeating number.  You use CNodeRef CConnMgr::FindNodeFromId(NodeId id) to get the CNode (and automatically AddRef(), btw) and use CNode::GetId() to get the id.  FindNodeFromId will return a null reference if the node has been deleted.  This is better than using CNode* since you are not holding up the node's cleanup when it disconnects.  You'll still need to periodically go thru your maps and clean out any entries whose nodeId doesn't match a live node in case a node disconnects in the middle of a mempool sync op.  Otherwise a connect, start sync, disconnect attack could fill your maps with entries.  However, this cleanup could happen rarely.

bissias replied:
Done. I went with NodeId and also added a cleanup method that is called from net.cpp.

gandrewstone commented on src/blockrelay/mempool_sync.h:
```diff
@@ -0,0 +1,188 @@
+// Use CVariableFastFilter if true, otherwise use CBloomFilter
+const bool COMPUTE_OPTIMIZED = true;
+
+class CMempoolSyncState
```
please add doxygen comments for all classes

bissias replied:
Done.

gandrewstone commented on src/blockrelay/mempool_sync.h:
```diff
@@ -0,0 +1,188 @@
+class CMempoolSyncInfo
+{
+public:
+    uint64_t nTxInMempool;
```
doxygen comment all public members

bissias replied:
Done.

gandrewstone commented on src/globals.cpp:
```diff
@@ -48,6 +49,8 @@
 
 #include <atomic>
 #include <boost/lexical_cast.hpp>
+#include <boost/thread/tss.hpp> // for boost::thread_specific_ptr
```
I don't see a use of thread_specific_ptr so remove this #include.  Also, in general don't use thread specific data in any project -- its unstable/difficult to use/link on some platforms.

bissias replied:
Weird. I don't remember adding this. Perhaps it's some rebase artifact. Anyhow, I removed it.

gandrewstone commented on src/globals.cpp:
```diff
@@ -265,6 +268,8 @@ CTweak<uint64_t> pruneIntervalTweak("prune.pruneInterval",
 
 CTweak<uint32_t> netMagic("net.magic", "network prefix override. If 0 use the default", 0);
 
+CTweak<uint32_t> randomlyDontInv("net.randomlyDontInv", "Skip sending an INV for some percent of transactions", 0);
```
should we qualify this with "for every node that supports mempool sync"?

bissias replied:
The way it is implemented now, any node can use the switch (although it is perhaps not a good idea if mempool sync is disabled). I can, however, make it available only to nodes that have enabled mempool sync and adjust the message accordingly.

gandrewstone commented on src/globals.cpp:
```diff
@@ -397,6 +402,26 @@ CTweak<uint64_t> grapheneFastFilterCompatibility("net.grapheneFastFilterCompatib
     "Support fast Bloom filter: 0 - either, 1 - fast only, 2 - regular only (default: either)",
     GRAPHENE_FAST_FILTER_SUPPORT);
 
+CTweak<bool> syncMempoolWithPeers("net.syncMempoolWithPeers", "Synchronize mempool with peers", false);
```
This is a periodic sync, right?  Maybe to give extra configurability, change to an integer "mempoolSyncPeriod" in seconds with 0 meaning "off".  Another thing that this has made me think of is that we should distribute sync evenly over time.  So if you have 10 nodes and the period is 1 minute, we should sync with 1 node every 6 seconds... this will minimize bandwidth.

bissias replied:
Overall, I'm in favor of adding the flexibility you're proposing. One concern I have is in allowing the requester to dictate the update frequency. This could be an attack vector. Currently the requester will not allow syncing more often than every 25 seconds from any given peer. I could of course add a default allowable sync frequency for the responder and add a sync frequency negotiation method similar to version negotiation. Do you think that's the way to go?

For your second point, let me make sure I'm understanding correctly. You're saying that whatever the negotiated sync frequency *per node* should be spread out so that you're syncing as frequently as possible while respecting the sync frequency of each peer. Is that right? Currently we randomly choose a peer at the (universal) sync frequency. This balances load, but does not maximize throughput.

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,503 @@
+    CMempoolSync tmp;
+    vRecv >> tmp;
+    auto mempoolSync = std::make_shared<CMempoolSync>(std::forward<CMempoolSync>(tmp));
+
```
You don't need to make this a shared pointer since you're not saving this to any map later.  This was probably some cut and paste code form graphene which needed sharted pointers for tracking blocks that were being reconstructed.

-vRecv >> tmp
-auto mempooSync ....
+vRecv >> mempoolSync

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,503 @@
+
+        // Sort out what hashes we have from the complete set of cheapHashes
+        for (size_t i = 0; i < blockCheapHashes.size(); i++)
+        {
```
nit: this could be a range based loop
-for (size_t i = 0; i < blockCheapHashes.size(); i++)
-uint64_t cheapHash = blockCheapHashes[i];
+for (uint64_t cheapHash : blockCheapHashes)

ptschip commented on src/blockrelay/mempool_sync.cpp:
```diff
@@ -0,0 +1,503 @@
+
+        // Skip if node is in IBD
+        if ((nCommonHeight < chainActive.Tip()->nHeight - 10) && (nSyncHeight < chainActive.Tip()->nHeight - 10))
+            continue;
```
maybe switch the 10 for DEFAULT_BLOCKS_FROM_TIP which we used in IsChainNearlySyncd(). I think this would be more in alignment with other areas of the code.

ptschip commented on src/net.cpp:
```diff
@@ -2278,6 +2283,17 @@ void ThreadMessageHandler()
+            // select node from whom to request mempool sync
+            CNode *syncPeer = SelectMempoolSyncPeer(vNodesCopy);
+            if (!(syncPeer == nullptr))
+                requester.RequestMempoolSync(syncPeer);
```
I don't think we need to start requesting mempool sync until our own chain is almost syncd and indeed we don't even request transactions until we're out of IBD anway.

-if (!(syncPeer == nullptr))
+if (!(syncPeer == nullptr) && IsChainNearlySyncd())

ptschip committed:
A few nits

ptschip committed:
Don't request a mempool sync unless our chain is nearly synced

We don't process INV's for transactions and then request transactions
before IBD is complete so there's no need to do any mempool syncing
either.

ptschip commented:
you have a formatting issue

ptschip commented:
Looks good now.  Why don't you squash all your commits down and then look at my PR that I opened against your repo?  Then i think we're done unless @gandrewstone has any other comments.

bissias commented:
> Looks good now. Why don't you squash all your commits down and then look at my PR that I opened against your repo? Then i think we're done unless @gandrewstone has any other comments.

@ptschip I still have at least one more thing to do for @gandrewstone, being the switch from `CNode*` to `nodeid` in `mempoolSyncRequested` and `mempoolSyncResponded`.

ptschip commented on src/net.cpp:
```diff
@@ -2291,6 +2292,10 @@ void ThreadMessageHandler()
+        // Periodically clear mempool sync maps in case peers have disconnected
+        if (GetStopwatchMicros() - lastMempoolSyncClear > MEMPOOLSYNC_CLEAR_FREQ_US)
+            ClearDisconnectedFromMempoolSyncMaps();
+
```
This clearing of mempoolsync maps should really be done in FinalizeNode() in main.cpp but I don't want to hold up the PR...we could fix this in a followup.

bissias replied:
I don't mind moving it but can you check [this comment](https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1879#discussion_r330789893) from @gandrewstone above? FinalizeNode acts on specific node (known to be disconnecting), he seems to be suggesting that nodes can "leak" in some other way?

ptschip committed:
Move the cleanup of the mempool sync maps to FinalizeNode()

This will reduce lock contention for vNodes.

ptschip commented:
Nodes can not leak that I'm aware of. FinalizeNode() is the very last thing that happens and is specifically used for cleaning up node information. There should be no possiblility of anything getting into the map for that nodeid once we get to finalizenode() since the node has already been erased from vNodes.

bissias committed:
Merge pull request #21 from ptschip/dev_sync

Cleanup sync maps in FinalizeNode()
